### ISSUE TITLE
Lines style migration problem

### ISSUE THREAD

**Issue by xavijam** [CartoDB/cartodb#9478]
We are migrating the stroke color of the line to the fill color.
<img width="366" alt="screen shot 2016-08-18 at 16 55 30" src="https://cloud.githubusercontent.com/assets/132146/17778526/990fa898-6564-11e6-9729-f438aead9642.png">

**Comment by juanignaciosl** [comment]
FYI this doesn't happen for actual _migrations_: creating a map with the Editor and then opening it with the Builder is fine:

![screen shot 2016-08-18 at 17 51 51](https://cloud.githubusercontent.com/assets/932968/17780651/7917167c-656c-11e6-9a55-8ca926242ef1.png)

I'll review why does it happen at direct creation from dashboard though.

**Comment by juanignaciosl** [comment]
It Works In My Machine™. It works in local development, not in production. Stay tuned 👻

**Comment by juanignaciosl** [comment]
Ok, `layersData` in production doesn't contain `style_properties`<= style migration is not being run.

**Comment by juanignaciosl** [comment]
Ok, after a talk with @xavijam we know what's happening here.

TL;DR: "recent" (post-builder-dataset-view-era) datasets canonical visualizations haven't `wizard_properties`, so maps based on them haven't (default) style properties.

*What happened before*

1. You import a dataset.
2. Editor Dataset page is open and `wizard_properties`are initialized at canonical visualization.
3. You create a map with that dataset. The map duplicates dataset canonical visualization properties, which have `wizard_properties`.
4. The layers get `style_properties` from `wizard_properties` migration when you first present the layers for the Builder.

*What happens now*
1. You import a dataset.
2. _Builder Dataset page hasn't anything related to styles, so styles are not initialized._
3. You create a map with that dataset. The map duplicates dataset canonical visualization properties, _which haven't `wizard_properties`_.

Solution alternatives:
a) Initialize styles at dashboard (front) on map creation. It's problematic because style initialization logic should be available not only at Builder but also at the dashboard.
b) Initialize styles at the Builder, when you load a visualization that hasn't `style_properties`. It's problematic because at the first load of a visualization it should initialize style if it hasn't, and make a PUT to sync them.
c) Initialize styles at the back, on map creation. It's problematic because we should code the logic that initializes based on default cartography and default cartocss in a similar fashion that front does, which is duplicated logic.


IMHO, the best tradeoff is b). @xavijam opinion is c).

cc @javisantana @xavijam @gfiorav

**Comment by xavijam** [comment]
The `c` option should not be that problematic, it would be a simple one. In fact I guess we are doing something similar for initialize new created layers.

**Comment by juanignaciosl** [comment]
@xavijam yep, I agree. My point about c isn't being complicated but not the best because of "partial duplication" (most creation logic will always be un front, picking some of it at back seems arbitrary. But I don't really love c either, both b and c have tradeoffs.

**Comment by gfiorav** [comment]
I'd go with b). 

IMO we need robustness over speed at this point in time (specially if it's a one time load). We can always to better later to trim a second of load time afterwards or redesign the styles stuff later on. It will take the same effort as doing it now (we're not making future compromises with b).

**Comment by xavijam** [comment]
Take into account that, if we create a map of 10 layers, we will need 10 `PUT` requests.

**Comment by gfiorav** [comment]
Yeah, but that's the nature of creating a 10 layer map: you get 10x the load 😄 

I mean, maybe I'm super off with this one but for me it's natural: not worth sacrificing our current robustness.

Besides, this is the type of thing that improves with technology. Everybody is trying to make 10 PUTs faster, but you can't expect them to solve your specific logic duplication issues. I feel it's more future proof to go with straightforward classic solution, unless it meas waaay to much load time (does it?).

**Comment by xavijam** [comment]
More than slowness, our main concern is make `PUT`s without any user action behind, I feel it is ugly.

**Comment by gfiorav** [comment]
I see what you mean.

It _is_ user action if you think of it. The user created a 10 layer map so here come 10 PUTs. I'm cool with any decision but let's really think if this is so bad.

**Comment by javisantana** [comment]
but B) means to do a PUT when opening a visualization. In the past we had a lot of problems, opening a visualization should not do anything. Only user actions should trigger PUT/POST/DELETE actions

For me c) makes sense and it's not that hard, it's only a template https://github.com/CartoDB/cartodb/blob/bc358f7b1c24f81abff554a277e357916af48b77/lib/assets/javascripts/cartodb3/data/camshaft-reference.js#L122

**Comment by juanignaciosl** [comment]
Some comments:

> Take into account that, if we create a map of 10 layers, we will need 10 PUT requests.

There aren't that many map creations with many tables at once, and even those shouldn't be a huge problem as it's only on first creation.

> I mean, maybe I'm super off with this one but for me it's natural: not worth sacrificing our current robustness.

I think the key of b is not robustness but maintainability. In fact, if robustness were the key I'd do 100% of the map initialization at backend, as it's a more controlled environment than a browser.

> More than slowness, our main concern is make PUTs without any user action behind, I feel it is ugly.

Certainly, that's true. It would be better on pure creation (either at back or front), not first initialization.

> For me c) makes sense and it's not that hard

Yes, it's not a matter of _hardness_ but _duplication_.

I'm not too convinced of any option. As I just had an small preference for b over c, IMHO, let's do c, but considering moving _more_ initialization to backend to avoid duplication of concerns. I'm just guessing but, for example, if front assigns a default CartoCSS on empty layer creation, that could be also deferred to back: if a layer creation contains CartoCSS backend should assume that it's not a default one so it should leave it as it is, but if it doesn't contain any, backend should initialize it. This way the initialization concern is only at back. Again, this is only an example, I hope that you get my point. How does it sound?

**Comment by gfiorav** [comment]
Agreed offline:

We will port some of the logic for setting default styles based on geometry to the backend. It should be noted that we still think there's no future separating responsibilities, but let's disagree and commit!

Some notes:

+ Style form definitions (JSON) changes should be notified to back people
+ Changes in style logic (we know use border radius or whatever) need to be maintained in back as well (fuel for our disagreement)

We all agree there's no beautiful solution that fits our schedule so we're cool with it :)
